### PR TITLE
[dynamo][compile-time] Caching VTs to reduce compile-time

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121031

Reduces the `torch.compile(backend="eager")` for this code

~~~
def fn(x):
    for _ in range(10000):
        # x = torch.sin(x)
        x = torch.ops.aten.sin(x)
        # x = sin(x)

    return x
~~~

From 18 seconds to 12 seconds.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

### CODE REVIEW

File: torch/_dynamo/variables/builder.py [hunk 1]
```diff
@@ -262,10 +262,17 @@ def __call__(self, value):
             if dup_guard:
                 self.install_guards(dup_guard)
             return side_effect_result
+
+        # TODO - Is id(value) sufficient?
+        # This is useful for LOAD_ATTR e.g. if we retrace export result, we will
+        # have each op with torch.ops.aten.
+        if id(value) in self.tx.cached_vts:
+            return self.tx.cached_vts[id(value)]
```
Comment (anijain2305) [inline torch/_dynamo/variables/builder.py:270]: This is what needs discussion.

**Review by jansel — COMMENTED** [review]
I think this idea could work.

**Comment by anijain2305** [comment]
@jansel this is ready for review.

File: torch/_dynamo/output_graph.py [hunk 2]
```diff
@@ -113,6 +113,40 @@
...
+
+    def add(self, value, source, vt):
+        key = VariableTrackerCacheKey(id(value), source)
+        self.cache[key] = vt
+
+    def clone(self):
+        # Needed for copy and restore graph state
+        new_cache = VariableTrackerCache()
+        for k, v in self.cache.items():
+            new_cache.cache[k] = v
```
Comment (jansel) [inline torch/_dynamo/output_graph.py:143]:
```suggestion
        new_cache.cache.update(self.cache)
```

**Review by jansel — APPROVED** [review]

**Pull request closed without merging**